src/Functions/toStartOfInterval.cpp

```diff
@@ -40,156 +325,228 @@ class FunctionToStartOfInterval : public IFunction
                 getName(), arguments.size());
         }
 
-        if (result_type_is_date)
-            return std::make_shared<DataTypeDate>();
-        else if (result_type_is_datetime)
-            return std::make_shared<DataTypeDateTime>(extractTimeZoneNameFromFunctionArguments(arguments, 2, 0, false));
-        else if (result_type_is_datetime_64)
+        switch (result_type)
         {
-            auto scale = 0;
-
-            if (interval_type->getKind() == IntervalKind::Nanosecond)
-                scale = 9;
-            else if (interval_type->getKind() == IntervalKind::Microsecond)
-                scale = 6;
-            else if (interval_type->getKind() == IntervalKind::Millisecond)
-                scale = 3;
-
-            return std::make_shared<DataTypeDateTime64>(scale, extractTimeZoneNameFromFunctionArguments(arguments, 2, 0, false));
+            case ResultType::Date:
+                return std::make_shared<DataTypeDate>();
+            case ResultType::DateTime:
+            {
+                const size_t time_zone_arg_num = (arguments.size() == 2 || (arguments.size() == 3 && third_argument == ThirdArgument::IsTimezone)) ? 2 : 3;
+                return std::make_shared<DataTypeDateTime>(extractTimeZoneNameFromFunctionArguments(arguments, time_zone_arg_num, 0, false));
+            }
+            case ResultType::DateTime64:
+            {
+                UInt32 scale = 0;
+                if (interval_type->getKind() == IntervalKind::Nanosecond)
+                    scale = 9;
+                else if (interval_type->getKind() == IntervalKind::Microsecond)
+                    scale = 6;
+                else if (interval_type->getKind() == IntervalKind::Millisecond)
+                    scale = 3;
+
+                const size_t time_zone_arg_num = (arguments.size() == 2 || (arguments.size() == 3 && third_argument == ThirdArgument::IsTimezone)) ? 2 : 3;
+                return std::make_shared<DataTypeDateTime64>(scale, extractTimeZoneNameFromFunctionArguments(arguments, time_zone_arg_num, 0, false));
+            }
         }
 
-        UNREACHABLE();
+        std::unreachable();
     }
 
     ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, size_t /* input_rows_count */) const override
     {
         const auto & time_column = arguments[0];
         const auto & interval_column = arguments[1];
-        const auto & time_zone = extractTimeZoneFromFunctionArguments(arguments, 2, 0);
-        auto result_column = dispatchForTimeColumn(time_column, interval_column, result_type, time_zone);
+
+        ColumnWithTypeAndName origin_column;
+        const bool has_origin_arg = (arguments.size() == 3 && isDateOrDate32OrDateTimeOrDateTime64(arguments[2].type)) || arguments.size() == 4;
+        if (has_origin_arg)
+            origin_column = arguments[2];
+
+        const size_t time_zone_arg_num = (arguments.size() == 3 && isString(arguments[2].type)) ? 2 : 3;
```

Comment:
Any particular reason why we are no longer checking for `(arguments.size() == 2`? The calculation is supposed to be the same as in l. 457 and l. 470.